README.md
# Callbacks-Framework-SMP
[![version][version_badge]][changelog]
[![CodeFactor][codefactor_badge]](https://www.codefactor.io/repository/github/regorxxx/Callbacks-Framework-SMP/overview/main)
[![CodacyBadge][codacy_badge]](https://www.codacy.com/gh/regorxxx/Callbacks-Framework-SMP/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=regorxxx/Callbacks-Framework-SMPP&amp;utm_campaign=Badge_Grade)
![GitHub](https://img.shields.io/github/license/regorxxx/Callbacks-Framework-SMP)  
A helper script for [Spider Monkey Panel](https://theqwertiest.github.io/foo_spider_monkey_panel) and [foobar2000](https://www.foobar2000.org) which allows to easily attach and manage event listeners associated to SMP callbacks without wrapping or repeating code.  A lifesaver. 

## The problem with current SMP callbacks
Complex scripts, like those with UI, usually have interactive items at multiple places which must be linked to a callback to perform actions at certain events. While SMP allows modularization using include(), callbacks must be placed within a main file since redefinition of the function simply overwrites the previous one. That clearly limits the ability to just mix modules and or add event listeners to non-final scripts.
![carbon(5)](https://user-images.githubusercontent.com/83307074/181040854-b05abb1d-c49e-42c8-83c2-94b847485a57.png)
![carbon(6)](https://user-images.githubusercontent.com/83307074/181041496-78448af2-8147-443e-a0f0-2f0109b71658.png)

Using this framework it would translate into this:

![carbon(7)](https://user-images.githubusercontent.com/83307074/181042039-6813117b-ea2c-485e-a131-86c7565fd695.png)

## Features
- Adds event listeners to SMP callbacks.
- Removes event listeners on demand.
- Allows to find if an event listener is active.
- Listeners are able to self-remove.
- Mimics [mozilla implementation](https://developer.mozilla.org/es/docs/Web/API/EventTarget/removeEventListener).
- Implements [SPM planned design](https://github.com/TheQwertiest/foo_spider_monkey_panel/projects/1).
- Bypasses common limitation of including files with existing callbacks, wrapping, etc.

## Usage
Just add an event listener to an specific event/callback.
```javascript
const listener = () => {console.log('Hello world!');}
addEventListener('on_mouse_lbtn_up', listener);
```

That's all. To remove it just do:
```javascript
removeEventListener('on_mouse_lbtn_up', listener);
```

Multiple listeners can be attached to the same event:
```javascript
addEventListener('on_mouse_lbtn_up', () => {console.log('1!');});
addEventListener('on_mouse_lbtn_up', () => {console.log('2!');});
addEventListener('on_mouse_lbtn_up', () => {console.log('3!');});
addEventListener('on_mouse_lbtn_up', () => {console.log('14!');});
```

Listener removal may be done using their UUID or function:
```javascript
const listener = () => {console.log('Hello world!');}
const uuid = addEventListener('on_mouse_lbtn_up', listener);
removeEventListener('on_mouse_lbtn_up', null, uuid);
```

There are more usage examples on the 'examples' folder.

## Installation
Since the framework only requires 1 file, i.e. the main one, you can simply include it along any other script where you usually need a callback.

![carbon(1)](https://user-images.githubusercontent.com/83307074/181011693-6f737dc4-0a86-4e2a-be2a-aa168128f1b1.png)

[changelog]: CHANGELOG.md
[version_badge]: https://img.shields.io/github/release/regorxxx/Callbacks-Framework-SMP.svg
[codacy_badge]: https://api.codacy.com/project/badge/Grade/3e59f8dccd204721a7801197d6c336ed
[codefactor_badge]: https://www.codefactor.io/repository/github/regorxxx/Callbacks-Framework-SMP/badge/main


CHANGELOG.md
# Changelog

## [Table of Contents]
- [Unreleased](#unreleased)
- [1.2.2](#122---2023-02-23)
- [1.2.1](#121---2022-10-01)
- [1.2.0](#120---2022-09-28)
- [1.1.0](#110---2022-08-12)
- [1.0.0](#100---2022-07-26)

## [Unreleased][]
### Added
### Changed
### Removed
### Fixed

## [1.2.2] - 2023-02-23
### Added
### Changed
### Removed
### Fixed
- Fix to on_mouse_rbtn_up disable default menu behavior.

## [1.2.1] - 2022-10-01
### Added
### Changed
### Removed
### Fixed
- Crash due to non defined this when adding a single listener to the stack.

## [1.2.0] - 2022-09-28
### Added
- parentWindow.eventListener: added new property to 'this' within event listeners scope. It may be used to remove itself after executing it once for ex. Contains name of the callback (event) and listener id.
- removeEventListenerSelf: shorthand to 'removeEventListener(this.eventListener.event, null, this.eventListener.id)' within event listeners scope.
### Changed
### Removed
### Fixed

## [1.1.0] - 2022-08-12
### Added
- removeEventListeners: new function 'removeEventListeners(event)' to remove all listeners attached to a callback. Event may be a string or array of strings with the name of callbacks to remove, i.e. 'on_playlists_changed' or  ['on_playlists_changed', 'on_script_unload'].
### Changed
### Removed
### Fixed

## [1.0.0] - 2022-07-26
### Added
- First release.
### Changed
### Removed
### Fixed

[Unreleased]: https://github.com/regorxxx/Callbacks-Framework-SMP/compare/v1.2.2...HEAD
[1.2.2]: https://github.com/regorxxx/Callbacks-Framework-SMP/compare/v1.2.1...v1.2.2
[1.2.1]: https://github.com/regorxxx/Callbacks-Framework-SMP/compare/v1.2.0...v1.2.1
[1.2.0]: https://github.com/regorxxx/Callbacks-Framework-SMP/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/regorxxx/Callbacks-Framework-SMP/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/regorxxx/Callbacks-Framework-SMP/compare/997dd95...v1.0.0

examples/01_callbacks.js
'use strict';
//26/07/22

include('..\\helpers\\callbacks_xxx.js');

/*
	Callbacks-XXX Framework example:
		Load file into a panel
		- L. Button click: Outputs to console log a message
		- R. Button click: Enables/disables L. button click functionality
*/

window.DefineScript('Callbacks-Framework-SMP');
fb.ShowPopupMessage('- L. Button click: Outputs to console log a message\n- R. Button click: Enables/disables L. button click functionality', window.ScriptInfo.Name);

// Event listener
const listener = () => {console.log('Hello world!');}

addEventListener('on_mouse_lbtn_up', listener);

addEventListener('on_mouse_rbtn_up', () => {
	console.log('Pressed R. Button');
	if (findEventListener('on_mouse_lbtn_up', listener) !== -1) {
		console.log('Disabled L. Button listener');
		removeEventListener('on_mouse_lbtn_up', listener);
		return true;
	} else {
		console.log('Enabled L. Button listener');
		addEventListener('on_mouse_lbtn_up', listener); 
		return true;
	}
});

helpers/callbacks_xxx.js
'use strict';
//31/12/22

/*
	Usage:
		- addEventListener(): to attach a new function to a callback. For ex: to display a text on console when using L. Button click
			const listener = () => {console.log('Hello world!');};
			const id = addEventListener('on_mouse_lbtn_up', listener);
		- removeEventListener(): to remove a previously added listener. For ex:
			removeEventListener('on_mouse_lbtn_up', listener); // Previous id may also be useds	
		- findEventListener(): to find if some listener is attached to a callback. For ex:
			const idx = findEventListener('on_mouse_lbtn_up', listener); // -1 since it has been previously removed
		- removeEventListeners(): to remove all listeners atached to a callback. For ex:
			removeEventListeners('on_focus'); // true
			
		Following [mozilla implementation](https://developer.mozilla.org/es/docs/Web/API/EventTarget/removeEventListener)
		event listeners may be removed using their UUID or event listener function itself.
		addEventListener() returns the UUID for the provided	function, which may be useful for anonymous or inline functions.
		removeEventListener() accepts both the listener (2nd arg) or UUID (3rd arg) as arguments.
		
		Callback registering is done under the hood automatically although it may be done manually using
		registerCallback() or registerAllCallbacks(). Note registering a callback multiple times will effectively duplicate
		the listener calls, so it should be avoided. Removing all listeners associated to a callback will not unregister
		the callback itself, which would be an undesirable effect in some cases and harmless in any other case.
		
		Callbacks previously added to any script will be wrapped and used along those added with event listeners as long as
		callback registering is done automatically, so they may work in conjunction. 
		
		See examples for more info.
*/

const callbacks = {
	on_always_on_top_changed:				{listeners: [], bRegistered: false},
	on_char: 								{listeners: [], bRegistered: false},
	on_colours_changed:						{listeners: [], bRegistered: false},
	on_cursor_follow_playback_changed:		{listeners: [], bRegistered: false},
	on_drag_drop:							{listeners: [], bRegistered: false},
	on_drag_enter:							{listeners: [], bRegistered: false},
	on_drag_leave:							{listeners: [], bRegistered: false},
	on_drag_over:							{listeners: [], bRegistered: false},
	on_dsp_preset_changed:					{listeners: [], bRegistered: false},
	on_focus:								{listeners: [], bRegistered: false},
	on_font_changed:						{listeners: [], bRegistered: false},
	on_get_album_art_done:					{listeners: [], bRegistered: false},
	on_item_focus_change:					{listeners: [], bRegistered: false},
	on_item_played:							{listeners: [], bRegistered: false},
	on_key_down:							{listeners: [], bRegistered: false},
	on_key_up:								{listeners: [], bRegistered: false},
	on_library_items_added:					{listeners: [], bRegistered: false},
	on_library_items_changed:				{listeners: [], bRegistered: false},
	on_library_items_removed:				{listeners: [], bRegistered: false},
	on_load_image_done:						{listeners: [], bRegistered: false},
	on_main_menu:							{listeners: [], bRegistered: false},
	on_main_menu_dynamic:					{listeners: [], bRegistered: false},
	on_metadb_changed:						{listeners: [], bRegistered: false},
	on_mouse_lbtn_dblclk:					{listeners: [], bRegistered: false},
	on_mouse_lbtn_down:						{listeners: [], bRegistered: false},
	on_mouse_lbtn_up:						{listeners: [], bRegistered: false},
	on_mouse_leave:							{listeners: [], bRegistered: false},
	on_mouse_mbtn_dblclk:					{listeners: [], bRegistered: false},
	on_mouse_mbtn_down:						{listeners: [], bRegistered: false},
	on_mouse_mbtn_up:						{listeners: [], bRegistered: false},
	on_mouse_move:							{listeners: [], bRegistered: false},
	on_mouse_rbtn_dblclk:					{listeners: [], bRegistered: false},
	on_mouse_rbtn_down:						{listeners: [], bRegistered: false},
	on_mouse_rbtn_up:						{listeners: [], bRegistered: false},
	on_mouse_wheel:							{listeners: [], bRegistered: false},
	on_mouse_wheel_h:						{listeners: [], bRegistered: false},
	on_notify_data:							{listeners: [], bRegistered: false},
	on_output_device_changed:				{listeners: [], bRegistered: false},
	on_paint:								{listeners: [], bRegistered: false},
	on_playback_dynamic_info:				{listeners: [], bRegistered: false},
	on_playback_dynamic_info_track:			{listeners: [], bRegistered: false},
	on_playback_edited:						{listeners: [], bRegistered: false},
	on_playback_follow_cursor_changed:		{listeners: [], bRegistered: false},
	on_playback_new_track:					{listeners: [], bRegistered: false},
	on_playback_order_changed:				{listeners: [], bRegistered: false},
	on_playback_pause:						{listeners: [], bRegistered: false},
	on_playback_queue_changed:				{listeners: [], bRegistered: false},
	on_playback_seek:						{listeners: [], bRegistered: false},
	on_playback_starting:					{listeners: [], bRegistered: false},
	on_playback_stop:						{listeners: [], bRegistered: false},
	on_playback_time:						{listeners: [], bRegistered: false},
	on_playlist_item_ensure_visible:		{listeners: [], bRegistered: false},
	on_playlist_items_added:				{listeners: [], bRegistered: false},
	on_playlist_items_removed:				{listeners: [], bRegistered: false},
	on_playlist_items_reordered:			{listeners: [], bRegistered: false},
	on_playlist_items_selection_change:		{listeners: [], bRegistered: false},
	on_playlist_stop_after_current_changed:	{listeners: [], bRegistered: false},
	on_playlist_switch:						{listeners: [], bRegistered: false},
	on_playlists_changed:					{listeners: [], bRegistered: false},
	on_replaygain_mode_changed:				{listeners: [], bRegistered: false},
	on_script_unload:						{listeners: [], bRegistered: false},
	on_selection_changed:					{listeners: [], bRegistered: false},
	on_size:								{listeners: [], bRegistered: false},
	on_volume_change:						{listeners: [], bRegistered: false}
}

const parentWindow = this; // This is Window in this context without SMP wrapping
parentWindow.eventListener = {event: null, id: null};

function addEventListener(event, listener, bRegister = true) {
	if (!callbacks.hasOwnProperty(event)) {console.log('addEventListener: event does not exist -> ' + event); return false;}
	const id = UUID();
	callbacks[event].listeners.push({id, listener});
	if (bRegister && !callbacks[event].bRegistered) {registerCallback(event);} // Only add those callbacks needed to the global context
	return {event, id};
}

function findEventListener(event, listener = null, id = null) {
	if (!callbacks.hasOwnProperty(event)) {return -1;}
	if (!listener && !id) {return -1;}
	const idx = callbacks[event].listeners.findIndex((event) => {
		return event.id === id || event.listener === listener;
	});
	return idx;
}

function removeEventListener(event, listener = null, id = null) {
	if (!callbacks.hasOwnProperty(event)) {return false;}
	if (!listener && !id) {return false;}
	const idx = findEventListener(event, listener, id);
	return idx !== -1 && callbacks[event].listeners.splice(idx, 1);
}

function removeEventListeners(event) {
	if (Array.isArray(event)) {
		event.forEach((ev) => {removeEventListeners(ev);});
	} else {
		if (!callbacks.hasOwnProperty(event)) {console.log('removeEventListeners: event does not exist -> ' + event); return false;}
		callbacks[event].listeners = [];
	}
	return true;
}

// Should only be called within an event listener, since 'this' points to 'parentWindow'
const removeEventListenerSelf = () => {return removeEventListener(this.eventListener.event, null, this.eventListener.id);}

/*
	Register callbacks
*/
const fireEvents = function(event) {
	return function() {
		let bReturn = event === 'on_mouse_rbtn_up' && callbacks[event].listeners.length; // To be used by on_mouse_rbtn_up to disable default menu
		callbacks[event].listeners.forEach((eventListener) => {
			if (typeof this === 'undefined') {console.log(event); console.log(eventListener.listener.toString());}
			this.eventListener = {event, id: eventListener.id};
			bReturn = eventListener.listener.apply(this, arguments);
			this.eventListener = {event: null, id: null};
		});
		return bReturn;
	};
}

function registerCallback(event) {
	if (typeof parentWindow[event] !== 'undefined') {
		const oldFunc = parentWindow[event];
		parentWindow[event] = function() {
			const cache = oldFunc.apply(parentWindow, arguments);
			const output = fireEvents(event).apply(parentWindow, arguments);
			return (cache || output); // To be used by on_mouse_rbtn_up to disable default menu
		}
	} else {
		parentWindow[event] = fireEvents(event).bind(parentWindow);
	}
	callbacks[event].bRegistered = true;
}

function registerAllCallbacks() {
	Object.keys(callbacks).forEach((event) => {registerCallback(event);});
}

/*
	Helpers
*/
if (typeof UUID === 'undefined') {
	var UUID = () => {
		return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace( /[xy]/g , function(c) {
				const rnd = Math.random() * 16 | 0, v = c === 'x' ? rnd : (rnd&0x3|0x8) ;
				return v.toString(16);
		});
	};
}

examples/02_callbacks_hell.js
'use strict';
//26/07/22

include('..\\helpers\\callbacks_xxx.js');

/*
	Callbacks-XXX Framework example:
		Load file into a panel
		- L. Button click: Outputs to console log a message
		- R. Button click: Enables/disables L. button click functionality
		
	How manual callbacks and event listeners may work in conjunction ...
*/

window.DefineScript('Callbacks-Framework-SMP');
fb.ShowPopupMessage('- L. Button click: Outputs to console log a message\n- R. Button click: Enables/disables L. button click functionality\n\nNote in this example the R. button switch only affects a part of the message.', window.ScriptInfo.Name);

// Standard callback
function on_mouse_lbtn_up() {
	console.log('Hey!');
}

// Event listener
const listener = () => {console.log('Hello world!');}

addEventListener('on_mouse_lbtn_up', listener);

addEventListener('on_mouse_rbtn_up', () => {
	console.log('Pressed R. Button');
	if (findEventListener('on_mouse_lbtn_up', listener) !== -1) {
		console.log('Disabled L. Button listener');
		removeEventListener('on_mouse_lbtn_up', listener);
		return true;
	} else {
		console.log('Enabled L. Button listener');
		addEventListener('on_mouse_lbtn_up', listener); 
		return true;
	}
});

// Wrapped callback when we are unsure other files add their own manually
// In this use-case, the previously defined callback will be wrapped
function on_mouse_lbtn_up_second() {
	console.log('Ho!');
}
if (typeof on_mouse_lbtn_up !== 'undefined') {
	const oldFunc = on_mouse_lbtn_up;
	on_mouse_lbtn_up = function() {
		oldFunc.apply(this, arguments);
		on_mouse_lbtn_up_second.apply(this, arguments);
	}
} else {
	on_mouse_lbtn_up = on_mouse_lbtn_up_second;
}